Mathlib/CategoryTheory/ConcreteCategory/Basic.lean

```diff
@@ -204,15 +227,196 @@ def HasForget₂.trans (C : Type u) [Category.{v} C] [ConcreteCategory.{w} C]
     φ.app Y (F.map f x) = G.map f (φ.app X x) := by
   simpa only [Functor.map_comp] using congr_fun ((forget D).congr_map (φ.naturality f)) x
 
+section ConcreteCategory
+
+/-- A concrete category is a category `C` where objects correspond to types and morphisms to
+(bundled) functions between those types.
+
+In other words, it has a fixed faithful functor `forget : C ⥤ Type`.
+
+Note that `ConcreteCategory` potentially depends on three independent universe levels,
+* the universe level `w` appearing in `forget : C ⥤ Type w`
+* the universe level `v` of the morphisms (i.e. we have a `Category.{v} C`)
+* the universe level `u` of the objects (i.e `C : Type u`)
+They are specified that order, to avoid unnecessary universe annotations.
+-/
+class ConcreteCategory (C : Type u) [Category.{v} C]
+    (FC : outParam <| C → C → Type*) {CC : outParam <| C → Type w}
+    [outParam <| ∀ X Y, FunLike (FC X Y) (CC X) (CC Y)] where
+  /-- Convert a morphism of `C` to a bundled function. -/
+  (hom : ∀ {X Y}, (X ⟶ Y) → FC X Y)
+  /-- Convert a bundled function to a morphism of `C`. -/
+  (ofHom : ∀ {X Y}, FC X Y → (X ⟶ Y))
+  (hom_ofHom : ∀ {X Y} (f : FC X Y), hom (ofHom f) = f := by aesop_cat)
+  (ofHom_hom : ∀ {X Y} (f : X ⟶ Y), ofHom (hom f) = f := by aesop_cat)
+  (id_apply : ∀ {X} (x : CC X), hom (𝟙 X) x = x := by aesop_cat)
+  (comp_apply : ∀ {X Y Z} (f : X ⟶ Y) (g : Y ⟶ Z) (x : CC X),
+    hom (f ≫ g) x = hom g (hom f x) := by aesop_cat)
+
+export ConcreteCategory (id_apply comp_apply)
+
+variable {C : Type u} [Category.{v} C] {FC : C → C → Type*} {CC : C → Type w}
+variable [∀ X Y, FunLike (FC X Y) (CC X) (CC Y)]
+variable [ConcreteCategory C FC]
+
+/-- `ToType X` converts the object `X` of the concrete category `C` to a type.
+
+This is an `abbrev` so that instances on `X` (e.g. `Ring`) do not need to be redeclared.
+-/
+abbrev ToType [ConcreteCategory C FC] := CC
+
+/-- `ToHom X Y` is the type of (bundled) functions between objects `X Y : C`.
+
+This is an `abbrev` so that instances (e.g. `RingHomClass`) do not need to be redeclared.
+-/
+abbrev ToHom [ConcreteCategory C FC] := FC
+
+namespace ConcreteCategory
+
+attribute [simp] id_apply comp_apply
+
+/-- We can apply morphisms of concrete categories by first casting them down
+to the base functions.
+-/
+instance {X Y : C} : CoeFun (X ⟶ Y) (fun _ ↦ ToType X → ToType Y) where
+  coe f := hom f
+
+/--
+`ConcreteCategory.hom` bundled as an `Equiv`.
+-/
+def homEquiv {X Y : C} : (X ⟶ Y) ≃ ToHom X Y where
+  toFun := hom
+  invFun := ofHom
+  left_inv := ofHom_hom
+  right_inv := hom_ofHom
+
+lemma hom_bijective {X Y : C} : Function.Bijective (hom : (X ⟶ Y) → ToHom X Y) :=
+  homEquiv.bijective
+
+lemma hom_injective {X Y : C} : Function.Injective (hom : (X ⟶ Y) → ToHom X Y) :=
+  hom_bijective.injective
+
+/-- In any concrete category, we can test equality of morphisms by pointwise evaluations. -/
+@[ext] lemma ext {X Y : C} {f g : X ⟶ Y} (h : hom f = hom g) : f = g :=
```

Comment:
This is a replacement for `hom_ext`, that currently is defined in terms of `HasForget`. (Similarly for `congr_fun`, `congr_arg`, `hom_id` and `hom_comp` below.) Since we have no `ConcreteCategory` instances yet, we'll unfortunately need to keep both spellings for a while. At some point we'll also have to decide which name is nicer :)